File: app/src/main/java/com/dominikgaller/alpinebooking/guide/outbound/persistence/write/GuideTourMapper.java
package com.dominikgaller.alpinebooking.guide.outbound.persistence.write;

import com.dominikgaller.alpinebooking.guide.core.domain.guidetour.GuideTour;
import com.dominikgaller.alpinebooking.guide.core.domain.guidetour.GuideTourId;
import com.dominikgaller.alpinebooking.guide.core.domain.guidetour.GuideTourStatus;
import com.dominikgaller.alpinebooking.shared.domain.TourId;
import com.dominikgaller.alpinebooking.jooq.tables.records.GuideTourRecord;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.Optional;
import java.util.UUID;

/**
 * Pure mapping utility: converts between the {@link GuideTour} aggregate and
 * the jOOQ {@link GuideTourRecord}.
 *
 * <p>No Spring dependency. No IO. {@code TIMESTAMP} columns are stored as UTC
 * {@link LocalDateTime} and converted to/from {@link Instant} using UTC offset.
 *
 * <p>SDD: See {@code documentation/use-cases/uc05-start-tour.spec.md}.
 */
class GuideTourMapper {

    GuideTourRecord toRecord(final GuideTour tour) {
        final GuideTourRecord record = new GuideTourRecord();
        record.setId(tour.id().value().toString());
        record.setTourId(tour.tourId().value());
        record.setScheduledStart(LocalDateTime.ofInstant(tour.scheduledStart(), ZoneOffset.UTC));
        record.setStatus(tour.status().name());
        record.setStartedAt(
                Optional.ofNullable(tour.startedAt())
                        .map(i -> LocalDateTime.ofInstant(i, ZoneOffset.UTC))
                        .orElse(null));
        return record;
    }

    GuideTour toDomain(final GuideTourRecord record) {
        final Instant startedAt = Optional.ofNullable(record.getStartedAt())
                .map(ldt -> ldt.toInstant(ZoneOffset.UTC))
                .orElse(null);
        return GuideTour.reconstitute(
                new GuideTourId(UUID.fromString(record.getId())),
                new TourId(record.getTourId()),
                record.getScheduledStart().toInstant(ZoneOffset.UTC),
                GuideTourStatus.valueOf(record.getStatus()),
                startedAt
        );
    }
}

File: app/src/main/kotlin/com/example/service/guide/outbound/persistence/GuideTourMapper.kt
package com.example.service.guide.outbound.persistence

import com.example.service.guide.core.domain.guidetour.GuideTour
import com.example.service.guide.core.domain.guidetour.GuideTourId
import com.example.service.guide.core.domain.guidetour.GuideTourStatus
import com.example.service.shared.domain.TourId
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.util.UUID

/**
 * Pure mapping between the [GuideTour] aggregate and [GuideTourJpaEntity].
 *
 * `Instant` is converted to and from `LocalDateTime` at UTC, matching the booking side. A
 * null `startedAt` stays null: a tour that has not started must remain distinguishable
 * from one that started at the epoch.
 */
internal class GuideTourMapper {
    fun toEntity(guideTour: GuideTour) =
        GuideTourJpaEntity(
            id = guideTour.id.value.toString(),
            tourId = guideTour.tourId.value,
            scheduledStart = LocalDateTime.ofInstant(guideTour.scheduledStart, ZoneOffset.UTC),
            status = guideTour.status.name,
            startedAt = guideTour.startedAt?.let { LocalDateTime.ofInstant(it, ZoneOffset.UTC) },
        )

    fun toDomain(entity: GuideTourJpaEntity) =
        GuideTour.reconstitute(
            id = GuideTourId(UUID.fromString(entity.id)),
            tourId = TourId(entity.tourId),
            scheduledStart = entity.scheduledStart.toInstant(ZoneOffset.UTC),
            status = GuideTourStatus.valueOf(entity.status),
            startedAt = entity.startedAt?.toInstant(ZoneOffset.UTC),
        )
}
